File: training-data/java/SubsetSum.java
public class SubsetSum {
    public static boolean hasSubsetSum(int[] nums, int target) {
        int n = nums.length;
        boolean[][] dp = new boolean[n + 1][target + 1];
        for (int i = 0; i <= n; i++) dp[i][0] = true;
        for (int i = 1; i <= n; i++) {
            for (int t = 1; t <= target; t++) {
                dp[i][t] = dp[i - 1][t];
                if (nums[i - 1] <= t) dp[i][t] = dp[i][t] || dp[i - 1][t - nums[i - 1]];
            }
        }
        return dp[n][target];
    }

    public static void main(String[] args) {
        System.out.println(hasSubsetSum(new int[] {3, 34, 4, 12, 5, 2}, 9));
    }
}

File: training-data/matlab/subset_sum.m
function found = subset_sum(nums, target)
    n = numel(nums);
    dp = false(n + 1, target + 1);
    dp(:, 1) = true;
    for i = 1:n
        for t = 1:target
            dp(i + 1, t + 1) = dp(i, t + 1);
            if nums(i) <= t
                dp(i + 1, t + 1) = dp(i + 1, t + 1) || dp(i, t - nums(i) + 1);
            end
        end
    end
    found = dp(n + 1, target + 1);
end

disp(subset_sum([3, 34, 4, 12, 5, 2], 9))
disp(subset_sum([1, 2, 3], 7))
